[CSS] Initial value for -webkit-line-clamp can't be expressed with its grammar

nt1m opened the pull request:
#### 25af638fc712d1eb8b3b20929a5c89189a460660
<pre>
[CSS] Initial value for -webkit-line-clamp can&apos;t be expressed with its grammar
<a href="https://bugs.webkit.org/show_bug.cgi?id=248638">https://bugs.webkit.org/show_bug.cgi?id=248638</a>
<a href="https://rdar.apple.com/103158259">rdar://103158259</a>

Reviewed by Darin Adler and Oriol Brufau.

`line-clamp: none` is a valid value according to: <a href="https://drafts.csswg.org/css-overflow-4/#propdef-line-clamp">https://drafts.csswg.org/css-overflow-4/#propdef-line-clamp</a>

It also represents the behavior of the initial value of not clamping the lines.

Parse this value, so the default behavior of `-webkit-line-clamp` can be expressed via a CSS value.

* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/webkit-line-clamp-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::fromCSSValue):
* Source/WebCore/css/CSSProperties.json:

Canonical link: <a href="https://commits.webkit.org/281826@main">https://commits.webkit.org/281826@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/839d39da7b1806f0169f55e52e895a383a34f475

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61164 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/40525 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/13745 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65114 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/11713 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63294 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48202 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/11988 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49453 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/8155 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/63198 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/37699 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/52991 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/30283 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/34379 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/64/builds/10216 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/10625 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/56196 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/10511 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/66845 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5111 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10325 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/56825 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5133 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/52954 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/57019 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/13634 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4231 "Passed tests") | | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/36329 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/37412 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/38506 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/37156 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSPrimitiveValueMappings.h:
```diff
@@ -144,6 +144,9 @@ template<> inline LineClampValue fromCSSValue(const CSSValue& value)
 {
     auto& primitiveValue = downcast<CSSPrimitiveValue>(value);
 
+    if (primitiveValue.isValueID() && primitiveValue.valueID() == CSSValueNone)
```
I believe there is no need to check is `isValueID` because `valueID` returns `CSSValueInvalid` in that case. I think we did that to make code like this easier to write.

darinadler replied:
Also, another way to do this is to assert the `valueID` is `CSSValueNone` instead of `ASSERT_NOT_REACHED` and not add any new code.

Loirooriol reviewed (approved):
LGTM with Darin's suggestions

nt1m committed:
[CSS] Initial value for -webkit-line-clamp can't be expressed with its grammar
https://bugs.webkit.org/show_bug.cgi?id=248638
rdar://103158259

Reviewed by Darin Adler and Oriol Brufau.

`line-clamp: none` is a valid value according to: https://drafts.csswg.org/css-overflow-4/#propdef-line-clamp

It also represents the behavior of the initial value of not clamping the lines.

Parse this value, so the default behavior of `-webkit-line-clamp` can be expressed via a CSS value.

* LayoutTests/imported/w3c/web-platform-tests/css/css-overflow/parsing/webkit-line-clamp-valid-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/cssom/cssom-getPropertyValue-common-checks-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
(WebCore::fromCSSValue):
* Source/WebCore/css/CSSProperties.json:

Canonical link: https://commits.webkit.org/281826@main
